allspice/ratelimiter.py

```diff
@@ -0,0 +1,66 @@
+import functools
+import time
+
+import requests
+
+
+class RateLimiter:
+    """
+    A decorator to rate limit a function call.
+
+    :param max_calls: Maximum number of calls per period
+    :param period: Time period in seconds
+
+    Example:
+
+        @RateLimiter(max_calls=10, period=1)  # 10 requests per second
+        def make_request(url):
+            pass
+    """
+
+    def __init__(self, max_calls, period=1.0):
+        assert period > 0, "Period should be a positive number"
+        assert max_calls >= 1, "max_calls should be at least one"
+        self.max_calls = int(max_calls)
+        self.period = float(period)
+        self.calls = 0
+        self.reset_time = time.time()
+
+    def __call__(self, func):
```

Comment:
Its unclear why we need two separate classes and wrapping of functions.

I think it would be more clear if we just override the request.Session request function and do the rate limiting wrapper logic in that.

While ratelimiter is much more general I think the trade off in clarity will be worth it.

Reply:
To clarify, you're suggesting dropping `RateLimiter` and subclassing `request.Session` with `RateLimitedSession` and just overriding the `request` method, right?

That makes sense, I'll make the change 👍

Reply:
Changed. The main reason I did it this way earlier was that it made it easier to unit test RateLimiter without relying on an external server, but later I ended up writing tests for the session as well anyway, so I think it doesn't make sense to keep the annotation around.